src/gc.h

```diff
@@ -177,28 +178,55 @@ typedef struct {
     char *data;
 } jl_gc_pagemeta_t;
 
-// Page layout:
-//  Newpage freelist: sizeof(void*)
-//  Padding: GC_PAGE_OFFSET - sizeof(void*)
-//  Blocks: osize * n
-//    Tag: sizeof(jl_taggedvalue_t)
-//    Data: <= osize - sizeof(jl_taggedvalue_t)
+typedef struct {
+    _Atomic(jl_gc_pagemeta_t *) page_metadata_back;
+} jl_gc_global_page_pool_t;
+
+extern jl_gc_global_page_pool_t global_page_pool_clean;
+extern jl_gc_global_page_pool_t global_page_pool_freed;
+
+#define GC_BACKOFF_MIN 4
+#define GC_BACKOFF_MAX 12
+
+STATIC_INLINE void gc_backoff(int *i) JL_NOTSAFEPOINT
+{
+    if (*i < GC_BACKOFF_MAX) {
+        (*i)++;
+    }
+    for (int j = 0; j < (1 << *i); j++) {
+        jl_cpu_pause();
+    }
+}
+
+// Lock-free stack implementation taken
+// from Herlihy's "The Art of Multiprocessor Programming"
+
+STATIC_INLINE void push_lf_page_metadata_back(jl_gc_global_page_pool_t *pool, jl_gc_pagemeta_t *elt) JL_NOTSAFEPOINT
+{
+    while (1) {
+        jl_gc_pagemeta_t *old_back = jl_atomic_load_relaxed(&pool->page_metadata_back);
+        elt->next = old_back;
+        if (jl_atomic_cmpswap(&pool->page_metadata_back, &old_back, elt)) {
+            break;
+        }
+        jl_cpu_pause();
+    }
+}
+
+STATIC_INLINE jl_gc_pagemeta_t *pop_lf_page_metadata_back(jl_gc_global_page_pool_t *pool) JL_NOTSAFEPOINT
+{
+    while (1) {
+        jl_gc_pagemeta_t *old_back = jl_atomic_load_relaxed(&pool->page_metadata_back);
+        if (old_back == NULL) {
+            return NULL;
+        }
+        if (jl_atomic_cmpswap(&pool->page_metadata_back, &old_back, old_back->next)) {
```

Comment:
To check my understanding: this is also not vulnerable to the ABA problem because we know that it is impossible that `push_lf_page_metadata_back` is called concurrently with it (so `old_back->next` cannot have changed in the meantime during this function). Is that right?

Reply:
`push_lf*` could be called concurrently with `pop_lf*`.

I believe this implementation would not be vulnerable to ABA because once a node is popped from one of the `page_pool_t` it can't be pushed to it again unless we are in the sweeping phase, which is itself serial.

Reply:
That seems reasonable, yeah